text/000-rustdoc-cfgs-handling.md

```diff
@@ -0,0 +1,281 @@
+
+It does not aim to allow having a same item with different `cfg`s to appear more than once in the generated documentation.
+
+It does not aim to document items which are *inactive* under the current configuration (i.e., “`cfg`'ed out”).
+
+# Motivation
+[motivation]: #motivation
+
+The goal of this RFC is to stabilize the possibility to add visual markers to the rendered documentation to know under which conditions an item is available.
+
+Providing this information to users will solve a common issue: “Why can I see this item in the documentation and yet can't use it in my code?”.
+The end goal being to provide this information automatically so that the documentation maintenance cost won't increase.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC proposes to add the following attributes:
+
+  * `#![doc(auto_cfg(enable))]`/`#[doc(auto_cfg(disable))]`
+
+    When this is turned on, `#[cfg]` attributes are shown in documentation just like `#[doc(cfg)]` attributes are. By default, `auto_cfg` will be enabled.
+
+  * `#[doc(cfg(...))]`
+
+    This attribute is used to document the operating systems, feature flags, and build profiles where an item is available. For example, `#[doc(cfg(unix))` will add a tag that says "this is supported on **unix** only" to the item.
+
+    The syntax of this attribute is the same as the syntax of the [`#[cfg(unix)]` attribute][cfg attribute] used for conditional compilation.
+
+  * `#![doc(cfg_hide(...))]` / `#[doc(cfg_show(...))]`
+
+    These attributes suppress or un-suppress the `auto_cfg` behavior for a particular configuration predicate.
+
+    For example, `#[doc(cfg_hide(windows))]` could be used in newer versions of the [`windows` crate] to prevent the "this is supported on **windows** only" tag from being shown on every single item.
+
+[cfg attribute]: https://doc.rust-lang.org/reference/conditional-compilation.html
+[`windows` crate]: https://docs.rs/windows/latest/windows/
+
+All of these attributes can be added to a module or to the crate root, and they will be inherited by the child items unless another attribute overrides it (except that `doc(cfg)` cannot be added to the crate root). This is why "opposite" attributes like `cfg_hide` and `cfg_show` are provided: they allow a child item to override its parent.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The attributes
+
+### `#[doc(auto_cfg(enable))]`/`#[doc(auto_cfg(disable))]`
+
+This is a crate-level attribute. By default, `#[doc(auto_cfg)]` is enabled at the crate-level. When it's enabled, Rustdoc will automatically display `cfg(...)` compatibility information as-if the same `#[doc(cfg(...))]` had been specified.
```

Comment:
Is there a reason why this has to be a crate-level attribute? I don't see why it couldn't be applied on a per-module basis, for example; I imagine this is mostly a limitation of the existing implementation. In that case, it might be worthwhile ensuring that uses outside the crate level are linted with a warning that says they currently have no use, but might in the future, similar to how the `coverage` attribute is used.

Reply:
You can see the discussion about why it's crate-level only [here](https://github.com/rust-lang/rfcs/pull/3631#discussion_r1595617210).

Reply:
We could add it to *Future Possibilities*.

Reply:
Very good point!